File: Linked List/2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

 class Solution {
  public:
      ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
          ListNode* dummy = new ListNode(0);
          ListNode* temp = dummy;
          int bonus = 0;
  
          while (l1 || l2 || bonus) {
              int value1 = l1 ? l1->val : 0;
              int value2 = l2 ? l2->val : 0;
              int sum = value1 + value2 + bonus;
              
              bonus = sum / 10;
              int value = sum % 10;
  
              temp->next = new ListNode(value);
              temp = temp->next;
  
              l1 = l1 ? l1->next : nullptr;
              l2 = l2 ? l2->next : nullptr;
          }
  
          return dummy->next;
      }
  };
  

File: Linked List/2. Add Two Numbers/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        bonus = 0
        dummy = ListNode(0)
        temp = dummy

        while l1 or l2 or bonus:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            sum_val = val1 + val2 + bonus

            bonus = sum_val // 10
            val = sum_val % 10

            temp.next = ListNode(val)
            temp = temp.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return dummy.next